---
title: Homework 7
author: Ansel George
output:
    pdf_document:
        highlight: tango
        latex_engine: xelatex
fontsize: 11pt
mainfont: Roboto
monofont: Ubuntu Mono
---

```{r, message=F}
library(dplyr)
library(tidyr)
library(readr)
library(ggplot2)
library(mvtnorm)

library(ConsensusClusterPlus)

set.seed(10)
```

# Problem 1:

*Instead of using the built-in `kmeans` function, implement your own k-means
algorithm from scratch. Run your algorithm multiple times on a simulated 2-D
dataset that is easy to visualize. Are your k-means results the same? To solve
this problem, implement consensus clustering by resampling 80% of the data and
perform k-means each time.*

```{r}
# Compute likelihoods for picking centroids from the data given a set of
# already picked centroids.
distance_probs <- function(distances, centroids_idx) {
  if (length(centroids_idx) == 1) {
    pairwise_distances <- distances[,centroids_idx]
  } else {
    pairwise_distances <- apply(distances[,centroids_idx], 1, min)
  }
  return(pairwise_distances/sum(pairwise_distances))
}

# Assign elements to a cluster given a set of centroids.
assign_cluster <- function(X, centroids) {
  k <- dim(centroids)[1]
  nr <- dim(X)[1]
  nc <- dim(X)[2]

  distances <- matrix(rep(0, nr*k), nrow=nr, ncol=k)
  for (i in 1:k) {
    distances[,i] <- apply(X, 1, f <- function(x) {sum((x - centroids[i,])^2)^.5})
  }
  return(apply(distances, 1, which.min))
}

# Generate a set of centroids given a dataset with assignments to k clusters.
update_centroids <- function(X, assignments, k) {
  nr <- dim(X)[1]
  nc <- dim(X)[2]

  centroids <- matrix(0, nrow=k, ncol=nc)
  for (i in 1:k) {
    tmp <- matrix(X[assignments == i], ncol=nc)

    # Need extra (inexhaustive) handling for when trying to compute a new
    # centroid when no points are assigned to it...
    if (length(tmp) == 0) {
      centroids[i,] <- matrix(0, ncol=nc)
    } else if (is.null(dim(tmp))) {
      centroids[i,] <- tmp
    } else {
      centroids[i,] <- apply(tmp, 2, mean)
    }
  }

  return(centroids)
}

# k-means++ implementation
kmeans_plusplus <- function(X, k=2, tol=.005) {
  nr <- dim(X)[1]
  nc <- dim(X)[2]

  # Pick initial centroids from data
  all_distances <- as.matrix(dist(X), nrow=nr)

  centroids <- matrix(rep(0, k*nc), ncol=nc)
  centroids_idx <- c()
  centroids_idx[1] <- sample(1:nr, 1)
  centroids[1,] <- X[centroids_idx[1],]

  for (i in 2:k) {
    probs <- distance_probs(all_distances, centroids_idx)
    centroids_idx <- append(centroids_idx, sample(1:nr, 1, prob=probs))
    centroids[i,] <- X[centroids_idx[i],]
  }

  # Do the k-means thing.
  error <- tol+1
  assignments <- rep(0, nr)
  counter <- 0
  status <- 0
  while(error > tol) {
    assignments <- assign_cluster(X, centroids)
    new_centroids <- update_centroids(X, assignments, k)
    error <- (sum(new_centroids - centroids)^2)^.5
    centroids <- new_centroids

    # Exit if it's not converging.
    if (counter > 100) {
      status <- 1
      break
    }
    counter <- counter + 1
  }

  return(list(centroids=centroids, assignments=assignments, steps=counter,
              status=status))
}
```

Generate a dataset with 3 highly visible clusters.

```{r}
N <- 200

mu1 <- c(0, 0)
Sigma1 <- matrix(c(.3, .1, .1, .2), nrow=2, ncol=2)
X1 <- rmvnorm(N, mean=mu1, sigma=Sigma1, method="chol")

mu2 <- c(5, 5)
Sigma2 <- matrix(c(.6, .3, .3, .4), nrow=2, ncol=2)
X2 <- rmvnorm(N, mean=mu2, sigma=Sigma2, method="chol")

mu3 <- c(0, 5)
Sigma3 <- matrix(c(.6, -.2, -.2, .2), nrow=2, ncol=2)
X3 <- rmvnorm(N, mean=mu3, sigma=Sigma3, method="chol")

X <- rbind(X1, X2, X3)

qplot(x=X[,1], y=X[,2])
```

Run k-means for 3 clusters.

```{r}
k <- 3
res <- kmeans_plusplus(X, k)
res$centroids
res$status # 0 is good
res$steps
```

Visualise the cluster assignments.

```{r}
d <- data.frame(cbind(X, res$assignments))
colnames(d) <- c("X", "Y", "cluster")
ggplot(d) + aes(x=X,y=Y,colour=factor(cluster)) + geom_point()
```

Now to do the consensus clustering:

```{r}
reps <- 100          # number of replicates for each k
K <- c(2, 3, 4, 5)   # check for set of k values

nr <- dim(X)[1]
nc <- dim(X)[2]
nr_sample <- .8 * nr # subsample 80% of the dataset

all_consensus <- list() # stores a list of consensus matrices for each k
for (k_idx in 1:length(K)) {
  # Preallocate matrices for storing counts:
  consensus <- matrix(0, nrow=dim(X)[1], ncol=dim(X)[1]) # pairs that are in the same clusters
  counts <- matrix(0, nrow=dim(X)[1], ncol=dim(X)[1]) # count of times pairs show up in samplings
  consensus_probs <- matrix(0, nrow=dim(X)[1], ncol=dim(X)[1]) # for storing counts/probabilities

  for (i in 1:reps) {
    idx <- sample(1:(N*3), floor(nr_sample))
    res <- kmeans_plusplus(X[idx,], K[k_idx])

    # Count the number of times pairs (for each cluster) are in the same
    # cluster.
    for (j in 1:(K[k_idx])) {
      tmp <- rep(0, dim(X)[1])
      tmp[idx[which(res$assignments == j)]] <- 1
      consensus <- consensus + tmp %*% t(tmp)
    }

    # Count the number of times any pair appears in the subsampled dataset.
    tmp <- rep(0, dim(X)[1])
    tmp[idx] <- 1
    counts <- counts + tmp %*% t(tmp)
  }

  consensus_probs <- consensus/counts
  consensus_probs <- consensus_probs[upper.tri(consensus_probs)] # the matrix is symmetric, so ignore the bottom triangle.
  all_consensus[[k_idx]] <- consensus_probs
}
```

Plot the CDF for the consensus index:

```{r}
ci <- seq(0, 1, .01)
cdfs <- matrix(0, nrow=length(ci), ncol=length(K))
nr <- dim(X)[1]

for (i in 1:length(K)) {
  cdfs[,i] <- sapply(ci,
                     f<-function(l) {
                       sum(all_consensus[[i]]<=l)/(nr*(nr-1)/2)
                     })
}

cdf_frame <- tbl_df(as.data.frame(cbind(ci, cdfs)))
colnames(cdf_frame) <- c("consensus_idx", paste("K=", K, sep=""))
cdf_frame %>%
  gather(K, CDF, -consensus_idx) %>%
  ggplot() + aes(x=consensus_idx, y=CDF, color=K) + geom_line()
```

The CDF plot shows that the most suitable cluster number is $k=3$. For that
value, the CDF is largely flat between 0 and 1, meaning that very few pairs of
elements are classified in different clusters.


# Problem 2:

*Follow the steps in Consensus Clustering Demo, perform consensus clustering on
a publicly available gene expression dataset, and report your findings.*

*For convenience, I uploaded a single-cell RNA-seq dataset of mouse cortical
cells (you can learn more about the data from [this
paper](https://www.nature.com/articles/nn.4216). The data is filtered and now
only contains the top 1000 most variable genes across 1809 cells. Another file
is included with cell labels generated by the paper, so that you can compare
your result with theirs. You can download the files here:
[gene_expression.zip](https://canvas.uchicago.edu/courses/21617/files/2459510/download?wrap=1).*


```{r}
metadata <- read_csv('data/sample_labels_GSE71585.csv')
all_data <- read_csv('data/top1k_logTPM_GSE71585.csv')
```

```{r}
cluster_data <- data.matrix(all_data %>% select(-X1))
```

```{r}
title <- "cell"
res <- ConsensusClusterPlus(cluster_data, maxK = 6, reps = 50, pItem = 0.8,
                            pFeature = 1, title = title, clusterAlg = "hc",
                            distance = "pearson", seed = 10, plot = "png")
                            # distance = "pearson", seed = 10, plot = "pdf")
```

```{r}
res[[2]][["consensusMatrix"]][1:5, 1:5]

#consensusTree - hclust object
res[[2]][["consensusTree"]]

#consensusClass - the sample classifications
res[[2]][["consensusClass"]][1:5]

icl = calcICL(res, title=title, plot="png")
# icl = calcICL(res, title=title, plot="pdf")
head(icl[["clusterConsensus"]])

icl[["itemConsensus"]][1:5, ]
```

Consensus cluster graphs:

![Consensus Heatmap Legend](cell/consensus001.png)

![Consensus k=2](cell/consensus002.png)

![Consensus k=3](cell/consensus003.png)

![Consensus k=4](cell/consensus004.png)

![Consensus k=5](cell/consensus005.png)

![Consensus k=6](cell/consensus006.png)

![Consensus CDF](cell/consensus007.png)

![Consensus Delta Area](cell/consensus008.png)

![Consensus Tracking Plot](cell/consensus009.png)


ICL graphs:

![Item Consensus (p1)](cell/icl001.png)

![Item Consensus (p2)](cell/icl002.png)

![Item Cluster Consensus](cell/icl003.png)
