---
title: Homework 4
author: Ansel George
output: pdf_document
---

# Problem 1

*Can you build a hash table for 3-tuples for the following sequences and search
for string "TAGCTAGCT"? Write down the pseudo code and then implement it.*

## Pseudocode

## Implementation

```{c, eval=F}
S1 = "GCTGCTGCTGCTAAACGTTTGGGGCAGTCGAT"
S2 = "GGTGCTCCAAGCTTTTGAGTCTGCTAGTGTCAACCCT"
S3 = "GTGGGCCCCCTAGCTAGCTAGCTGGGGCAC"
S4 = "TGTCGCTGGCTGGACTGCTGATCGTAGTAG"
```

See code.

## Substring search

See code.


# Problem 2

*Implement Burrows-Wheeler Transform of "DOGWORD" and obtain its LF mapping.
How can we search for "GWO" using FM index (implementation is optional, just
write down the procedure)?*

See code.


# Problem 3

*Give an $O(n \log k)$-time algorithm to merge $k$ sorted lists into one sorted
list, where $n$ is the total number of elements in all the input lists. (Hint:
Use a min-heap for $k$-way merging.)*


# Problem 4

*Implement genome assembly using the De Bruijn graph approach following these
steps:*

## (i) Simulate the genome
*Sample a DNA sequence of length 1000 bp. You could just use uniform
distribution over the four nucleotides ATCG.*

```{r}
nucleotides <- c("A", "T", "C" ,"G")
dna <- sample(nucleotides, 1000, replace=T)
dna <- paste(unlist(dna), collapse='')
```

## (ii) Simulate read data
*Suppose the read length is 25 bp, sample 400 reads from the genome,
with uniform start position.*

```{r}
read_length <- 25
read_start <- sample(seq(1, 1000-25), 20, replace=T)
reads <- substring(dna, first=read_start, last=read_start+read_length-1)
```


## (iii) Construct a graph of $k$-mers with $k = 10$.


## (iv) Find the Eulerian path of the graph.
*Ideally, you solve this by writing your own code. It is acceptable if you use
some package with graph functions to solve this.*


## (v) Report the statistics of your result
*Comparing with the true genome sequences. For example, you can do an alignment
of the two sequences and report the percentage of identity.*


# Problem 5

We extend PCST algorithm in class to find disease related gene modules:

Instead of having an unweighted network, we now have co-expression gene
network. In this undirected network, any two genes are linked with an edge,
with the edge weight being the correlation of gene expression (-1 to 1) across
a given set of gene expression datasets. We assume that genes with high
co-expression (both positive and negative) are likely to be involved in the
same biological processes.

Given the disease relevance of genes (p-values) and the weighted co-expression
network,

## a)
Extend the PCST algorithm discussed in class to find disease related gene
modules. We have two criteria for choosing gene modules: they should be
enriched with genes with high disease relevance (small p-values), and the genes
in the subgraphs should be tightly co-expressed.

## b)
In PCST, we combine vertex and edge scores by choosing a relative weight that
balances the two types of evidence. For example, if we put too much weight on
edge scores, we may find a subnetwork with many co-expressed genes but no
disease relevance. Discuss how you may choose this relative weight.

## c)
PCST only returns the best tree. But for many complex disease, one would expect
to have multiple disease related gene modules. How would you address this
problem?
